### ISSUE TITLE
Integration with Capacitor

### ISSUE THREAD

**Issue by Saveriu** [authts/oidc-client-ts#537]
Hi!
I' trying to use this lib with Capacitor, any idea of how to override the popup open event for authentication?
I want to use the Capacitor Browser plugin instead of window.open.

Thanks!

**Comment by Badisi** [comment]
You should be able to do it with your own Navigator implementation.
You can have a look at [PopupNavigator](https://github.com/authts/oidc-client-ts/blob/1d8254343748c367ef46e0992c510759ece990a2/src/navigators/PopupNavigator.ts) and [UserManager.signinPopup](https://github.com/authts/oidc-client-ts/blob/12024b48738cde94de48ea37252f02811534f531/src/UserManager.ts#L177) to see how it's done for the popup mode.

You can also have a look at [BrowserTabNavigator](https://github.com/Badisi/auth-js/blob/main/projects/auth-js/oidc/mobile/browser-tab/BrowserTabNavigator.ts) and [BrowserTab](https://github.com/Badisi/auth-js/blob/main/projects/auth-js/oidc/mobile/browser-tab/BrowserTab.ts), which is how I was doing it with the previous `oidc-client-js` library. Unfortunately I had no time to migrate it to `oidc-client-ts` yet but should be able to do it next week if you wishes to use my library 😉

**Comment by Saveriu** [comment]
Thanks for your reply!
In fact, for now, I'm using oidc-client-js because I'm using the vuex-oidc wrapper for 3 of my apps.
Now I know that it will no longer be maintenained, I will certainly switch for another lib and do the wrapper myself. 
I could use your lib for that! 
I will check your code thanks again! Do you have be able to do the redirecting from the dialog to the app with a custom scheme?

**Comment by Badisi** [comment]
Yes, code for the custom url scheme callback is: [here](https://github.com/Badisi/auth-js/blob/main/projects/auth-js/oidc/oidc-auth-manager.ts#L439)

**Comment by Saveriu** [comment]
Hi again!
I'm developping a Capacitor App that requires silentRefresh and multi autority servers (the app can have multiple OidcClient).
I'm facing the choice of using your lib or ionic-appauth.
Ionic appauth doesn't provide silentRefresh and I needed to complete the gap. I also have some issue to check if the user is really authenticated (ie: a expired access token in the store is considered ok :/)
For a long time I used oidc-client.js so I think it will be more faisable to use yours.
For that, I'm developping a custom Navigator with @capacitor/browser and StateStorage with @capacitor/storage.
I could provide you my code if you want to integrate Capacitor capabilities when I'm done.

**Comment by pamapa** [comment]
It would be very nice for others to see how this feature can be implemented.

**Comment by Saveriu** [comment]
I'm reading the code right now and I see that the Navigator implementation is fixed inside the UserManager constructor.
I will have to create a new CapacitorUserManager to be able to customize the Navigator. 
I will be not really clean but it can be a starting point.

**Comment by WillooWisp** [comment]
@Saveriu did you manage to get this working using Capacitor on Android and iOS? If so, any sample would be much appreciated.

**Comment by Saveriu** [comment]
@WillooWisp , I didn't have much time lately but when I'm done I will send you the sample!

**Comment by danielrie** [comment]
> @WillooWisp , I didn't have much time lately but when I'm done I will send you the sample!

May you send me the sample, too? I am also struggling to get capacitor working with this lib.

**Comment by danielrie** [comment]
> @WillooWisp , I didn't have much time lately but when I'm done I will send you the sample!

May you send me the sample, too? I am also struggling to get capacitor working with this lib.

**Comment by danielrie** [comment]
@Saveriu May you send me the sample, too? I am also struggling to get capacitor working with this lib.

**Comment by Saveriu** [comment]
Hi, for my project, I switched for ionic-appauth. The issue that I mentionned earlier have been fixed.
But I will try to do it with this lib when I have some time!

**Comment by Badisi** [comment]
For anyone interested, I've finally managed to find time to do it 🎉

If you are looking for a `sample` : I had to extend the [UserManager](https://github.com/Badisi/auth-js/blob/main/projects/auth-js/oidc/user-manager.ts) class to add mobile capabilities and implement a custom [Navigator](https://github.com/Badisi/auth-js/blob/main/projects/auth-js/oidc/mobile/mobile-navigator.ts), [Window](https://github.com/Badisi/auth-js/blob/main/projects/auth-js/oidc/mobile/mobile-window.ts) and [Storage](https://github.com/Badisi/auth-js/blob/main/projects/auth-js/oidc/mobile/mobile-storage.ts) class.

---

If you are looking for a `ready to use solution` : my [library](https://github.com/Badisi/auth-js) is made to work with a single implementation for both desktop and mobile and actually support VanillaJS and Angular projects. Feel free to test it or to ask questions (as the documentation is yet to be written).

**Comment by Saveriu** [comment]
Nice! I didn't have time to do it myself, thank you!

**Comment by Annouar** [comment]
Hey @Badisi,

Nice work here :)
I was wondering if we could make a PR to accept a custom `handle` to oidc-client-ts login/logout functions. I don't know if it is a better idea than create a new class that extends `UserManager` and create methods for mobile the use the exact same logic, but with a custom `IWindow` object.

What do you think about it ?

**Comment by Canuckaholic** [comment]
From digging into the source code it looks like a custom `redirectNavigator` (instance of RedirectNavigator class) can be passed in when creating a new UserManager. In this custom redirectNavigator you can use Capacitor's `Browser.open({ url });` instead of the default `window.location`. Is this not encouraged though (since it's not documented anywhere that I could see)? Any reason to not expose the RedirectNavigator class?

**Comment by Excel1** [comment]
@Canuckaholic did you got it working? If so can you share how?

**Comment by hmoazzem** [comment]
I've been trying to get it working, and eventually came up with https://github.com/edgeflare/ngx-oidc which works straight-forward.

But for apps requiring regulatory compliance, the CapacitorStateStore (https://github.com/edgeflare/ngx-oidc/blob/main/ngx-oidc-capacitor/src/lib/capacitor-state-store.ts) implemented with `@capacitor/preferences` functioning much like `window.localStorage` isn't sufficient. In such cases, implement a store with something like https://github.com/martinkasa/capacitor-secure-storage-plugin.
